platform/pulpcore/app/models/content.py

```diff
@@ -3,10 +3,44 @@
 """
 import hashlib
 
+from django.conf import settings
 from django.db import models
 
 from pulpcore.app.models import Model, MasterModel, Notes, GenericKeyValueRelation
-from pulpcore.app.models.storage import ArtifactLocation
+
+
+class Artifact(Model):
+    """
+    A file associated with a piece of content.
+
+    Fields:
+
+        file (models.FileField): The stored file.
+        downloaded (models.BooleanField): The associated file has been successfully downloaded.
+        requested (models.BooleanField): The associated file has been requested by a client at
+            least once.
+        size (models.IntegerField): The size of the file in bytes.
+        md5 (models.CharField): The MD5 checksum of the file.
+        sha1 (models.CharField): The SHA-1 checksum of the file.
+        sha224 (models.CharField): The SHA-224 checksum of the file.
+        sha256 (models.CharField): The SHA-256 checksum of the file.
+        sha384 (models.CharField): The SHA-384 checksum of the file.
+        sha512 (models.CharField): The SHA-512 checksum of the file.
+    """
+    def storage_path(self, name):
+        digest = self.sha256
+        return "%s/artifacts/%s/%s" % (settings.MEDIA_ROOT, digest[0:2], digest[2:])
+
+    file = models.FileField(db_index=True, upload_to=storage_path, max_length=255)
+    downloaded = models.BooleanField(db_index=True, default=False)
+    requested = models.BooleanField(db_index=True, default=False)
+    size = models.IntegerField(blank=True, null=True)
+    md5 = models.CharField(max_length=32, blank=True, null=True)
+    sha1 = models.CharField(max_length=40, blank=True, null=True)
+    sha224 = models.CharField(max_length=56, blank=True, null=True)
+    sha256 = models.CharField(max_length=64, blank=True, null=True)
+    sha384 = models.CharField(max_length=96, blank=True, null=True)
+    sha512 = models.CharField(max_length=128, blank=True, null=True)
```

Comment:
I assume an artifact cannot be created without at least 1 digest, correct?
And, the query to find an existing artifact is by matching one of the digests?

like:
```
where md5 = <md5> or,
      sha1 = <sha1> or,
      sha224 = <sha224> or,
      sha256 = <sha256> or,
      ...
```
only for non-null digests.

If so, won't we need indexes on all of the digests?